File: src/systems/MapSystem.ts
// P6: world map + fast travel. Owns the discovered points of interest (the
// dungeon entrance, Eldric's camp, the town centre) and the persisted
// fast-travel unlock. Discovery is proximity-based; fast travel unlocks once
// the P6 onboarding quest is complete.
import type { DungeonSystem } from "./DungeonSystem";
import type { QuestSystem } from "./QuestSystem";
import { GRID_CHUNK } from "../world/Grid";

export interface PoiInfo {
  id: string;
  name: string;
  icon: string;
  x: number;
  y: number;
  discovered: boolean;
  /** P6b: boss lairs render with a distinct (red) marker. */
  boss?: boolean;
  /**
   * Visible without being "discovered".
   *
   * Discovery is a persisted list of ids, which suits fixed landmarks. A clue's
   * dig site is neither fixed nor something you find by exploring — the scroll
   * just told you where it is — so it is shown directly instead of being written
   * into the save's discovered list, which would fill up with dead sites.
   */
  always?: boolean;
}

/** The player-owned subset of map state (persisted with the save). */
export interface MapStore {
  discovered: string[];
  fastTravel: boolean;
  /** P6.1: tile indices the player has walked (for the coverage meter). */
  explored: number[];
}

export interface MapCoverage {
  pct: number; // 0..100 of tiles explored
  coarse: string[]; // one "E"/"." per chunk block, row-major
}

export interface MapSnapshot {
  size: number;
  player: { x: number; y: number };
  unlocked: boolean;
  pois: PoiInfo[];
  coverage: MapCoverage;
}

const POI_DISCOVER_RADIUS = 3; // tiles — step within this to "explore" a spot

export class MapSystem {
  private dungeon: DungeonSystem;
  private quest: QuestSystem;
  private size: number;
  private store: MapStore;
  /** P6b: resolves the Forest Ogre's current lair tile (its home). */
  private getOgre: () => { x: number; y: number } | null;
  private getClueSite: () => { x: number; y: number } | null = () => null;
  /** P6.1: per-tile exploration flags (indexed y * size + x). */
  private explored = new Uint8Array(0);

  constructor(
    size: number,
    dungeon: DungeonSystem,
    quest: QuestSystem,
    store: MapStore,
    getOgre: () => { x: number; y: number } | null = () => null,
    /** The active clue's dig site, if a hunt is running. */
    getClueSite: () => { x: number; y: number } | null = () => null
  ) {
    this.size = size;
    this.dungeon = dungeon;
    this.quest = quest;
    this.store = store;
    this.getOgre = getOgre;
    this.getClueSite = getClueSite;
    if (!this.store.discovered.includes("town")) this.store.discovered.push("town");
    this.store.explored ??= [];
    this.explored = new Uint8Array(size * size);
    for (const idx of this.store.explored) if (idx >= 0 && idx < this.explored.length) this.explored[idx] = 1;
  }

  private pois(): Omit<PoiInfo, "discovered">[] {
    const cx = Math.floor(this.size / 2);
    const base: Omit<PoiInfo, "discovered">[] = [
      { id: "town", name: "Isoperia Centre", icon: "🏠", x: cx, y: cx },
      { id: "caves", name: "The Caves", icon: "🕳️", x: this.dungeon.entrance.x, y: this.dungeon.entrance.y },
      { id: "eldric", name: "Eldric's Camp", icon: "🧭", x: this.quest.guide.x, y: this.quest.guide.y },
    ];
    const ogre = this.getOgre();
    if (ogre && !base.some((p) => p.x === ogre.x && p.y === ogre.y)) {
      // Skip the boss marker if it would sit exactly on another waypoint.
      base.push({ id: "ogre", name: "The Forest Ogre", icon: "👹", x: ogre.x, y: ogre.y, boss: true });
    }
    // The dig site is the only marker that moves, so it is appended last and
    // carries a fixed id — the discovery list must not fill up with dead sites.
    const dig = this.getClueSite();
    if (dig) base.push({ id: "clue_site", name: "Dig here", icon: "📜", x: dig.x, y: dig.y, always: true });
    return base;
  }

  get unlocked(): boolean { return this.store.fastTravel; }

  unlockFastTravel(): void {
    if (this.store.fastTravel) return;
    this.store.fastTravel = true;
  }

  snapshot(px: number, py: number): MapSnapshot {
    return {
      size: this.size,
      player: { x: px, y: py },
      unlocked: this.store.fastTravel,
      pois: this.pois().map((p) => ({ ...p, discovered: p.always === true || this.store.discovered.includes(p.id) })),
      coverage: this.coverage(),
    };
  }

  /** P6.1: mark the player's tile + 4-neighbours as walked (persisted). */
  recordExplore(px: number, py: number): void {
    const n = this.size;
    if (px < 0 || py < 0 || px >= n || py >= n) return;
    const mark = (x: number, y: number) => {
      if (x < 0 || y < 0 || x >= n || y >= n) return;
      const idx = y * n + x;
      if (this.explored[idx]) return;
      this.explored[idx] = 1;
      this.store.explored.push(idx);
    };
    mark(px, py); mark(px + 1, py); mark(px - 1, py); mark(px, py + 1); mark(px, py - 1);
  }

  /** P6.1: coverage % + a coarse per-chunk exploration mask for the UI. */
  private coverage(): MapCoverage {
    const n = this.size;
    let seen = 0;
    for (let i = 0; i < this.explored.length; i++) if (this.explored[i]) seen++;
    const cols = Math.max(1, Math.ceil(n / GRID_CHUNK));
    const coarse: string[] = [];
    for (let cb = 0; cb < cols; cb++) {
      for (let ca = 0; ca < cols; ca++) {
        let any = false;
        for (let y = cb * GRID_CHUNK; y < Math.min(n, (cb + 1) * GRID_CHUNK); y++) {
          for (let x = ca * GRID_CHUNK; x < Math.min(n, (ca + 1) * GRID_CHUNK); x++) {
            if (this.explored[y * n + x]) { any = true; break; }
          }
          if (any) break;
        }
        coarse.push(any ? "E" : ".");
      }
    }
    return { pct: Math.round((seen / (n * n)) * 1000) / 10, coarse };
  }

  /** Mark newly-explored POIs the player just walked near. Returns ids. */
  checkDiscoveries(px: number, py: number): string[] {
    const fresh: string[] = [];
    for (const p of this.pois()) {
      if (this.store.discovered.includes(p.id)) continue;
      if (Math.abs(p.x - px) + Math.abs(p.y - py) <= POI_DISCOVER_RADIUS) {
        this.store.discovered.push(p.id);
        fresh.push(p.id);
      }
    }
    return fresh;
  }

  /** Fast-travel target for a discovered waypoint, or null (locked/unknown). */
  travelTarget(id: string): { x: number; y: number } | null {
    if (!this.store.fastTravel) return null;
    const poi = this.pois().find((p) => p.id === id);
    if (!poi || !this.store.discovered.includes(id)) return null;
    return { x: poi.x, y: poi.y };
  }

  poiName(id: string): string {
    return this.pois().find((p) => p.id === id)?.name ?? id;
  }
}

File: unity/Assets/Isoperia/Core/Runtime/Systems/MapSystem.cs
using System;
using System.Collections.Generic;
using Isoperia.Core.State;
using Isoperia.Core.World;

namespace Isoperia.Core.Systems
{
    /// <summary>A point of interest on the world map.</summary>
    public sealed class Poi
    {
        public string Id;
        public string Name;
        public string Icon;
        public int X;
        public int Y;
        public bool Boss;

        /// <summary>
        /// Visible without being "discovered".
        ///
        /// Discovery is a persisted list of ids, which suits fixed landmarks. A
        /// clue's dig site is neither fixed nor something you find by exploring —
        /// the scroll already told you where it is — so it shows directly rather
        /// than being written into the save.
        /// </summary>
        public bool Always;

        public bool Discovered;
    }

    /// <summary>
    /// World map, discovery and fast travel. Port of
    /// <c>src/systems/MapSystem.ts</c>.
    ///
    /// The dungeon entrance, the guide's camp, the boss lair and the clue dig
    /// site are all supplied as lookups rather than by holding references to
    /// those systems. The TypeScript already did that for two of them; doing it
    /// for all four means this file has no dependency on systems that are still
    /// being ported, and it keeps the map from becoming the place everything
    /// else gets wired through.
    /// </summary>
    public sealed class MapSystem
    {
        /// <summary>Manhattan distance within which walking discovers a landmark.</summary>
        public const int DiscoverRadius = 3;

        private readonly int _size;
        private readonly PlayerState _player;

        private readonly Func<(int X, int Y)?> _dungeonEntrance;
        private readonly Func<(int X, int Y)?> _guideCamp;
        private readonly Func<(int X, int Y)?> _ogreLair;
        private readonly Func<(int X, int Y)?> _clueSite;

        /// <summary>
        /// Per-tile flags, rebuilt from the persisted index list on construction.
        ///
        /// The save stores a LIST of indices while this is a dense array. The
        /// array exists so recordExplore can reject a repeat in O(1) — without it
        /// the list would grow without bound as the player walked back and forth
        /// over the same tiles, and on a 126x126 map that list is already the
        /// largest thing in the save.
        /// </summary>
        private readonly bool[] _explored;

        public MapSystem(int size, PlayerState player,
                         Func<(int X, int Y)?> dungeonEntrance = null,
                         Func<(int X, int Y)?> guideCamp = null,
                         Func<(int X, int Y)?> ogreLair = null,
                         Func<(int X, int Y)?> clueSite = null)
        {
            if (size <= 0) throw new ArgumentOutOfRangeException(nameof(size));

            _size = size;
            _player = player ?? throw new ArgumentNullException(nameof(player));

            _dungeonEntrance = dungeonEntrance ?? (() => null);
            _guideCamp = guideCamp ?? (() => null);
            _ogreLair = ogreLair ?? (() => null);
            _clueSite = clueSite ?? (() => null);

            // The town is known from the start; the player is standing in it.
            if (!_player.MapDiscovered.Contains("town")) _player.MapDiscovered.Add("town");

            _explored = new bool[size * size];
            foreach (double idx in _player.MapExplored)
            {
                int i = (int)idx;
                if (i >= 0 && i < _explored.Length) _explored[i] = true;
            }
        }

        public bool FastTravelUnlocked => _player.MapFastTravel;

        public void UnlockFastTravel() => _player.MapFastTravel = true;

        /// <summary>
        /// The landmarks, in a fixed order.
        ///
        /// The boss marker is skipped when it would land exactly on another
        /// waypoint, and the dig site is appended last with a fixed id.
        /// </summary>
        public List<Poi> Pois()
        {
            int c = _size / 2;

            var list = new List<Poi>
            {
                new Poi { Id = "town", Name = "Isoperia Centre", Icon = "🏠", X = c, Y = c },
            };

            (int X, int Y)? caves = _dungeonEntrance();
            if (caves.HasValue)
                list.Add(new Poi { Id = "caves", Name = "The Caves", Icon = "🕳️", X = caves.Value.X, Y = caves.Value.Y });

            (int X, int Y)? guide = _guideCamp();
            if (guide.HasValue)
                list.Add(new Poi { Id = "eldric", Name = "Eldric's Camp", Icon = "🧭", X = guide.Value.X, Y = guide.Value.Y });

            (int X, int Y)? ogre = _ogreLair();
            if (ogre.HasValue)
            {
                bool clash = false;
                foreach (Poi p in list) if (p.X == ogre.Value.X && p.Y == ogre.Value.Y) { clash = true; break; }
                if (!clash)
                    list.Add(new Poi { Id = "ogre", Name = "The Forest Ogre", Icon = "👹", X = ogre.Value.X, Y = ogre.Value.Y, Boss = true });
            }

            (int X, int Y)? dig = _clueSite();
            if (dig.HasValue)
                list.Add(new Poi { Id = "clue_site", Name = "Dig here", Icon = "📜", X = dig.Value.X, Y = dig.Value.Y, Always = true });

            foreach (Poi p in list)
                p.Discovered = p.Always || _player.MapDiscovered.Contains(p.Id);

            return list;
        }

        /// <summary>
        /// Mark the player's tile and its four neighbours as walked.
        ///
        /// Only NEW tiles are appended to the persisted list, so walking in
        /// circles does not grow the save.
        /// </summary>
        public void RecordExplore(int px, int py)
        {
            if (px < 0 || py < 0 || px >= _size || py >= _size) return;

            Mark(px, py);
            Mark(px + 1, py);
            Mark(px - 1, py);
            Mark(px, py + 1);
            Mark(px, py - 1);
        }

        private void Mark(int x, int y)
        {
            if (x < 0 || y < 0 || x >= _size || y >= _size) return;

            int idx = y * _size + x;
            if (_explored[idx]) return;

            _explored[idx] = true;
            _player.MapExplored.Add(idx);
        }

        public bool IsExplored(int x, int y) =>
            x >= 0 && y >= 0 && x < _size && y < _size && _explored[y * _size + x];

        /// <summary>Percentage of the world walked, to one decimal place.</summary>
        public double CoveragePercent()
        {
            int seen = 0;
            for (int i = 0; i < _explored.Length; i++) if (_explored[i]) seen++;

            return Math.Round(seen / (double)(_size * _size) * 1000, MidpointRounding.AwayFromZero) / 10;
        }

        /// <summary>
        /// One flag per chunk block, row-major: has the player been anywhere in
        /// it? This is what the map panel draws rather than 15,876 tiles.
        /// </summary>
        public List<bool> CoarseCoverage()
        {
            int cols = Math.Max(1, (int)Math.Ceiling(_size / (double)Grid.GridChunk));
            var coarse = new List<bool>(cols * cols);

            for (int cb = 0; cb < cols; cb++)
            {
                for (int ca = 0; ca < cols; ca++)
                {
                    bool any = false;

                    for (int y = cb * Grid.GridChunk; y < Math.Min(_size, (cb + 1) * Grid.GridChunk) && !any; y++)
                        for (int x = ca * Grid.GridChunk; x < Math.Min(_size, (ca + 1) * Grid.GridChunk); x++)
                            if (_explored[y * _size + x]) { any = true; break; }

                    coarse.Add(any);
                }
            }

            return coarse;
        }

        /// <summary>
        /// Discover any landmark the player has walked near. Returns the ids that
        /// flipped, so the caller can toast them once.
        /// </summary>
        public List<string> CheckDiscoveries(int px, int py)
        {
            var fresh = new List<string>();

            foreach (Poi p in Pois())
            {
                if (_player.MapDiscovered.Contains(p.Id)) continue;
                if (Math.Abs(p.X - px) + Math.Abs(p.Y - py) > DiscoverRadius) continue;

                _player.MapDiscovered.Add(p.Id);
                fresh.Add(p.Id);
            }

            return fresh;
        }

        /// <summary>
        /// Where a waypoint sends you, or null when fast travel is locked, the
        /// id is unknown, or the place has not been found yet.
        /// </summary>
        public (int X, int Y)? TravelTarget(string id)
        {
            if (!_player.MapFastTravel) return null;
            if (!_player.MapDiscovered.Contains(id)) return null;

            foreach (Poi p in Pois()) if (p.Id == id) return (p.X, p.Y);
            return null;
        }

        public string PoiName(string id)
        {
            foreach (Poi p in Pois()) if (p.Id == id) return p.Name;
            return id;
        }
    }
}
